pydomcfg/domzgr/zco.py

```diff
@@ -0,0 +1,280 @@
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ppa1: float
+            Coeff. controlling distibution of vertical levels
+            (default = pp_to_be_computed, i.e. computed from
+            ppdzmin, pphmax, ppkth and ppacr)
+        ldbletanh: bool
+            Logical flag to use or not double tanh stretching function
+            (default = False)
+        ppa2: float
+            Double tanh stretching function parameter
+        ppkth2: float
+            Double tanh stretching function parameter
+        ppacr2: float
+            Double tanh stretching function parameter
+        ln_e3_dep: bool
+            Logical flag to comp. e3 as fin. diff. (True) or
+            analyt. (False) (default = True)
+
+        Returns
+        -------
+        Dataset
+           Describing the 3D geometry of the model
+
+        Raises
+        -------
+        ValueError
+            If ldbletanh = True and parametrs are equal to 0.
+
+        """
+        self._ppdzmin = ppdzmin
+        self._pphmax = pphmax
+        self._ppkth = ppkth
+        self._ppacr = ppacr
+        self._ppsur = ppsur
+        self._ppa0 = ppa0
+        self._ppa1 = ppa1
+        self._ldbletanh = ldbletanh
+        self._ppa2 = ppa2
+        self._ppkth2 = ppkth2
+        self._ppacr2 = ppacr2
+        self._ln_e3_dep = ln_e3_dep
+
+        ds = self._init_ds()
+
+        # computing coeff. if needed
+        self._ppsur, self._ppa0, self._ppa1 = self._compute_pp()
+
+        # compute sigma-coordinates for z3 computation
+        kindx = DataArray(range(self._jpk), coords={"z": range(self._jpk)}, dims="z")
+        self._sigT = -self._sigma(kindx, "T")
+        self._sigW = -self._sigma(kindx, "W")
+        self._sigTp1 = -self._sigma(kindx + 1.0, "T")
+        self._sigWp1 = -self._sigma(kindx + 1.0, "W")
+
+        # compute z3 depths of vertical levels
+        for k in range(self._jpk):
+
+            if self._ppkth * self._ppacr == 0.0:
+                # uniform zco grid
+                suT = self._sigT[{"z": k}]
+                suW = self._sigW[{"z": k}]
+                s1T = s1W = s2T = s2W = 0.0
+                a1 = a3 = a4 = 0.0
+                a2 = self._pphmax
+            else:
+                # stretched zco grid
+                suT = self._sigTp1[{"z": k}]
+                suW = self._sigWp1[{"z": k}]
+                s1T = self._stretch_zco(self._sigT[{"z": k}])
+                s1W = self._stretch_zco(self._sigW[{"z": k}])
+                a1 = self._ppsur
+                a2 = self._ppa0 * (self._jpk - 1)
+                a3 = self._ppa1 * self._ppacr
+                # double tahh
+                if self._ldbletanh:
+                    if self._ppa2 * self._ppkth2 * self._ppacr2 == 0.0:
+                        raise ValueError(
+                            "ppa2, ppkth2 and ppacr2 must > 0. " "when ldbletanh = True"
+                        )
```

Comment:
Let's create a separate module to run at the very beginning that validates all arguments passed to `__call__()`?
We should raise errors as soon as possible.

Reply:
Agreed!